Windows MSVC build error:  C2065: 'M_PI': undeclared identifier

Greetings! 

As we are moving to package Ompl 2.0.0 (and also 1.7.0) in Conan Center, via PR https://github.com/conan-io/conan-center-index/pull/29277 (kudos @aleph-ra), during the build on Windows CI, the following error was found:

```
FAILED: src/ompl/CMakeFiles/ompl.dir/geometric/planners/rrt/src/ATRRT.cpp.obj 
C:\PROGRA~1\MICROS~2\2022\COMMUN~1\VC\Tools\MSVC\1444~1.352\bin\Hostx64\x64\cl.exe  /nologo /TP -DBOOST_ALL_DYN_LINK -DBOOST_ALL_NO_LIB -DBOOST_MATH_NO_LONG_DOUBLE_MATH_FUNCTIONS -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_TEST_DYN_LINK -D_HAS_AUTO_PTR_ETC=0 -Dompl_EXPORTS -IC:\j\29277-3\b\ompla9595ee3e1a6d\b\src\src -IC:\j\29277-3\b\ompla9595ee3e1a6d\b\build\Release\src -IC:\j\29277-3\boostfb4d3612ca900\p\include -IC:\j\29277-3\eigena71b2913e1aba\p\include\eigen3 /DWIN32 /D_WINDOWS /W3 /GR /EHsc /EHsc /MP /W1 -DNOMINMAX /O2 /Ob2 /DNDEBUG -MD -std:c++17 /showIncludes /Fosrc\ompl\CMakeFiles\ompl.dir\geometric\planners\rrt\src\ATRRT.cpp.obj /Fdsrc\ompl\CMakeFiles\ompl.dir\ /FS -c C:\j\29277-3\b\ompla9595ee3e1a6d\b\src\src\ompl\geometric\planners\rrt\src\ATRRT.cpp
cl : Command line warning D9025 : overriding '/W3' with '/W1'
C:\j\29277-3\b\ompla9595ee3e1a6d\b\src\src\ompl\geometric\planners\rrt\src\ATRRT.cpp(220): error C2065: 'M_PI': undeclared identifier
```

The full build log can be found here: https://c3i.jfrog.io/artifactory/cci-build-logs/cci/prod/PR-29277/3/package_build_logs/build_log_ompl_2_0_0_f1928c9fdbffd8f612cddaf0a73e5b8e_fb70c3537e304fe53ffa983eaa2e340bb756de2f.txt

The CI build result can be found here, so you can also check Mac and Linux: https://github.com/conan-io/conan-center-index/pull/29277/checks?check_run_id=74976798992

Locally, doing the same build I can reproduce the case, so it's not something isolated in the CI only.
You can take a look in my full build log as well: [ompl-2.0.0-windows-msvc194-cpp17-release-static.log](https://github.com/user-attachments/files/27517898/ompl-2.0.0-windows-msvc194-cpp17-release-static.log)

---- 

I understand `M_PI` does not work on Windows, because it's POSIX. Still, Microsoft offers the precompiler definition [_USE_MATH_DEFINES ](https://learn.microsoft.com/en-us/cpp/c-runtime-library/math-constants?view=msvc-170) to workaround this. Still there a few alternatives:

It's true C++20 comes with [std::numbers::pi](https://en.cppreference.com/cpp/numeric/constants), but I think is too much pushing the project to C++20 only this.

On the other hand, you are consuming Boost Math, which has [boost::math::pi](https://github.com/boostorg/math/blob/boost-1.91.0/include/boost/math/constants/constants.hpp#L273) and should be compatible

Feel free to ask more details about the case.

#### Environment
OMPL Version: 2.0.0
OS: Windows 10
Compiler: Visual Studio MSVC 19.44.35207.1
Architecture: x86_64
Cross-building: No
Build Type: Release

#### Steps to Reproduce:

Using a minimal build configuration to avoid mixing with external dependencies:

```
wget https://github.com/ompl/ompl/releases/download/2.0.0/ompl-2.0.0-Source.tar.gz
extract & go inside

cmake -S . -B build -G "Ninja" -DOMPL_BUILD_SHARED="OFF" -DOMPL_BUILD_PYTHON_BINDINGS="OFF" -DOMPL_BUILD_DEMOS="OFF" -DOMPL_BUILD_TESTS="OFF" -DOMPL_BUILD_VAMP="OFF" -DOMPL_VERSIONED_INSTALL="OFF" -DCMAKE_DISABLE_FIND_PACKAGE_Python="ON" -DCMAKE_DISABLE_FIND_PACKAGE_yaml-cpp="ON" -DCMAKE_DISABLE_FIND_PACKAGE_Doxygen="ON" -DCMAKE_DISABLE_FIND_PACKAGE_spot="ON" -DCMAKE_DISABLE_FIND_PACKAGE_Triangle="ON" -DCMAKE_DISABLE_FIND_PACKAGE_flann="ON" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" -DCMAKE_C_COMPILER="cl" -DCMAKE_CXX_COMPILER="cl" 

cmake --build build/
```

--- a/demos/PlanarManipulator/PlanarManipulator.cpp
+++ b/demos/PlanarManipulator/PlanarManipulator.cpp
@@ -193,7 +193,7 @@ void PlanarManipulator::FK(const double *joints, Eigen::Affine2d &eeFrame) const
 // Returns false if no solution exists to the given pose.
 bool PlanarManipulator::IK(std::vector<double> &solution, const Eigen::Affine2d &eeFrame) const
 {
-    std::vector<double> seed(numLinks_, M_PI / 2.0);
+    std::vector<double> seed(numLinks_, boost::math::constants::pi<double>() / 2.0);
     return IK(solution, seed, eeFrame);
 }
 
@@ -295,10 +295,10 @@ bool PlanarManipulator::IK(std::vector<double> &solution, const std::vector<doub
     for (size_t i = 0; i < solution.size(); ++i)
     {
         double angle = joints(i);
-        while (angle > M_PI)
-            angle -= 2.0 * M_PI;
-        while (angle < -M_PI)
-            angle += 2.0 * M_PI;
+        while (angle > boost::math::constants::pi<double>())
+            angle -= 2.0 * boost::math::constants::pi<double>();
+        while (angle < -boost::math::constants::pi<double>())
+            angle += 2.0 * boost::math::constants::pi<double>();
         solution[i] = angle;
     }
 
@@ -308,7 +308,7 @@ bool PlanarManipulator::IK(std::vector<double> &solution, const std::vector<doub
 bool PlanarManipulator::FABRIK(std::vector<double> &solution, const Eigen::Affine2d &eeFrame, double xyTol,
                                double thetaTol) const
 {
-    std::vector<double> seed(numLinks_, M_PI / 2.0);
+    std::vector<double> seed(numLinks_, boost::math::constants::pi<double>() / 2.0);
     return FABRIK(solution, seed, eeFrame, xyTol, thetaTol);
 }
 

--- a/demos/PlanarManipulator/PlanarManipulatorDemo.cpp
+++ b/demos/PlanarManipulator/PlanarManipulatorDemo.cpp
@@ -35,6 +35,7 @@
 /* Author: Ryan Luna */
 
 #include <fstream>
+#include <boost/math/constants/constants.hpp>
 
 #include "boost/program_options.hpp"
 #include "PolyWorld.h"
@@ -119,8 +120,8 @@ ompl::geometric::SimpleSetupPtr setupOMPL(Problem &problem)
     // Create the state space for the manipulator.
     ompl::base::StateSpacePtr space(new PlanarManipulatorStateSpace(numLinks));
     ompl::base::RealVectorBounds bounds(numLinks);
-    bounds.setLow(-M_PI);
-    bounds.setHigh(M_PI);
+    bounds.setLow(-boost::math::constants::pi<double>());
+    bounds.setHigh(boost::math::constants::pi<double>());
 
     // Bound the joints of the manipulator between [-PI, PI]
     space->as<PlanarManipulatorStateSpace>()->setBounds(bounds);

--- a/demos/PlanarManipulator/PlanarManipulatorPolyWorld.h
+++ b/demos/PlanarManipulator/PlanarManipulatorPolyWorld.h
@@ -41,6 +41,7 @@
 #include <Eigen/Dense>
 #include <cmath>
 #include <vector>
+#include <boost/math/constants/constants.hpp>
 
 // Create the 'corridor' environment for a unit-length planar manipulator with
 // n links.  The chain must navigate a narrow gap.
@@ -54,7 +55,7 @@ PolyWorld createCorridorProblem(int n, Eigen::Affine2d &basePose, Eigen::Affine2
     PolyWorld world("corridor", {minX, maxX}, {minY, maxY});
 
     const double len = 1.0 / n;
-    const double gap = len * (1.1 * M_PI + log(n) / (double)n);
+    const double gap = len * (1.1 * boost::math::constants::pi<double>() + log(n) / (double)n);
 
     std::vector<Point> pts(4);
 

--- a/demos/PlanarManipulator/PlanarManipulatorTSRRTConfig.h
+++ b/demos/PlanarManipulator/PlanarManipulatorTSRRTConfig.h
@@ -87,7 +87,7 @@ public:
 
         std::vector<double> lifted_angles;
         if (!manip_->FABRIK(lifted_angles, seed_angles, end_frame, /*xyTol=*/1e-4,
-                            /*thetaTol=*/2.0 * M_PI))
+                            /*thetaTol=*/2.0 * boost::math::constants::pi<double>()))
         {
             return false;
         }

--- a/demos/multilevel/MultiLevelPlanarManipulatorDemo.cpp
+++ b/demos/multilevel/MultiLevelPlanarManipulatorDemo.cpp
@@ -59,6 +59,7 @@
 
 #include <ompl/multilevel/datastructures/Projection.h>
 #include <ompl/multilevel/datastructures/projections/SE2_R2.h>
+#include <boost/math/constants/constants.hpp>
 
 using namespace ompl::geometric;
 
@@ -136,8 +137,8 @@ int main()
     // #########################################################################
     ompl::base::StateSpacePtr space(new PlanarManipulatorStateSpace(numLinks));
     ompl::base::RealVectorBounds bounds(numLinks);
-    bounds.setLow(-M_PI);
-    bounds.setHigh(M_PI);
+    bounds.setLow(-boost::math::constants::pi<double>());
+    bounds.setHigh(boost::math::constants::pi<double>());
     space->as<PlanarManipulatorStateSpace>()->setBounds(bounds);
     manipulator.setBounds(bounds.low, bounds.high);
 

--- a/src/ompl/geometric/planners/rrt/src/ATRRT.cpp
+++ b/src/ompl/geometric/planners/rrt/src/ATRRT.cpp
@@ -40,6 +40,7 @@
 #include "ompl/tools/config/SelfConfig.h"
 #include "ompl/tools/config/MagicConstants.h"
 #include <limits>
+#include <boost/math/constants/constants.hpp>
 
 ompl::geometric::ATRRT::ATRRT(const base::SpaceInformationPtr &si) : base::Planner(si, "ATRRT")
 {
@@ -217,7 +218,7 @@ ompl::base::PlannerStatus ompl::geometric::ATRRT::solve(const base::PlannerTermi
     double approxDifference = std::numeric_limits<double>::infinity();
     // UsefulCycles constant
     double volumeFreeSpace = 0.9 * si_->getMaximumExtent();
-    double volumeUnitBall2D = M_PI;
+    double volumeUnitBall2D = boost::math::constants::pi<double>();
     unsigned int d = si_->getStateDimension();
     gamma_ = 2.5 * 2 * std::pow((1.0 + 1.0 / d), 1.0 / d) * std::pow(volumeFreeSpace / volumeUnitBall2D, 1.0 / d);
 

